URL and Title changes trigger history entry updates

andrewcarlotti opened the pull request:
This is a change to the way that history entries are generated, to fix known issues where the history is incomplete (#5012 and #2835) or has incorrect page titles. Both forms of misbehaviour can currently be observed on YouTube (when clicking through several different videos).

This replaces the existing (different) algorithms for webengine and webkit with a common (almost) algorithm, which works roughly as follows. Whenever the page title changes and the page url does not, then the new title is saved in the history entry in place of the old one. A new history entry is created whenever the url changes and either the title also changes, or a `loadFinished` (webengine) or `initialLayoutCompleted` (webkit) signal is received. (These different signals are the ones currently used; I don't see any reason for this not to be unified). There are also some subtleties concerning recognising redirects.

To implement this change it is necessary to amend history entries after initially adding them to the database. (Having logged the signals emitted when browsing YouTube, this retrospective editing of history seems to be unavoidable). This has two consequences. Firstly, updates to history entries assume that the pair (atime, url) is sufficiently unique that it is safe to update all entries with the same atime and url simultaneously. There are some convoluted edge-cases in which this would leave history entries with an inaccurate title, but I don't think we should worry about those.

The second consequence is that more of the history entry processing has to be undertaken on a per-tab basis. This required some refactoring of the code, including changing the signature of add_url. As a consequence, several of the tests are now broken. I haven't yet attempted to fix them, as I thought it would be better to first have a discussion/review of the main code changes. The refactoring also entirely replaces the changes made in #4699, where consecutive history entries were deduplicated on a global basis. I think it is more appropriate to carry out this deduplication on a per-tab basis, as I have done here.


Some questions/considerations for futher development:
1. There are some cases in which it would be desirable to save a new history entry even without a loadFinished signal or a change in the page title. One such example would be when viewing consecutively two YouTube videos with the same title. On the other hand, without the signal or title change check there would be a lot of history entries created when panning many common online maps (including Google Maps). I think the first situation is sufficiently unusual that it is ok to miss the entry there, when weighed against the benefit of not having an excessive quantity of map history entries, but I'd be interested to hear any other opinions.

2. Should reloading a page create a new entry? Currently the behaviour depends on whether there have been any page loads in other tabs. With this change, reloading would never create a new entry. There's an argument to be made that it should at least sometimes create a new entry - either at all times, or whenever the url has changed since the last history entry (from that tab). Either of these would help mitigate the issue in the first question.

3. A more technical question. The `_on_history_trigger` method for webengine previously tried calling `self._widget.page()`, apparently to check whether the tab had been destroyed. I don't see why this would be necessary, and have removed that check while refactoring; is that likely to reintroduce some subtle bug, or was that guard no longer necessary?

toofar commented:
I haven't looked at the code in any detail but your description sounds reasonable. It's a pity that everything is based around signals so we can't get the ID of the row back when doing inserts which would allow doing more precise updates.

> A more technical question. The _on_history_trigger method for webengine previously tried calling self._widget.page(), apparently to check whether the tab had been destroyed. I don't see why this would be necessary, and have removed that check while refactoring; is that likely to reintroduce some subtle bug, or was that guard no longer necessary?

This kind of thing happens when signals get queued and don't make it to listeners before something else happens, like closing a tab. Those checks have unfortunately been necessary all over the place, both Qt and the Chromium bits of WebEngine have their own interthread/process communication and event loops. Probably safer to leave it in.

> Don't save the title if it's generated from the URL

Personally I would prefer entries had a title, even if it were the same as the URL, over an empty one.

andrewcarlotti commented:
> > A more technical question. The _on_history_trigger method for webengine previously tried calling self._widget.page(), apparently to check whether the tab had been destroyed. I don't see why this would be necessary, and have removed that check while refactoring; is that likely to reintroduce some subtle bug, or was that guard no longer necessary?
> 
> This kind of thing happens when signals get queued and don't make it to listeners before something else happens, like closing a tab. Those checks have unfortunately been necessary all over the place, both Qt and the Chromium bits of WebEngine have their own interthread/process communication and event loops. Probably safer to leave it in.

Ah, that makes sense. I've restored the check (in a different place - it's now called for Webkit as well).


> > Don't save the title if it's generated from the URL
> 
> Personally I would prefer entries had a title, even if it were the same as the URL, over an empty one.

I was just reproducing existing behaviour (for WebEngine) here (originally introduced in dfed2f9c); this change does introduce the same behaviour for WebKit.

rcorre commented on qutebrowser/misc/sql.py:
```diff
@@ -371,6 +371,34 @@ def insert_batch(self, values, replace=False):
+        values = {
+            "filter" + key: value for (key, value) in filter_values.items()}
+        values.update({
+            "new"+key: value for (key, value) in new_values.items()})
```
Having to concat strings here to make the keys match is a bit nasty, but I don't have a better way to do it right now :man_shrugging:

andrewcarlotti commented on qutebrowser/browser/browsertab.py:
```diff
@@ -1041,9 +1047,76 @@ def _update_load_status(self, ok: bool) -> None:
+        # Don't save the title if it's generated from the URL
+        title = self.title()
+        title_url = QUrl(url)
+        title_url.setScheme('')
```
Good question - I don't think it is. I just transferred the existing code from webenginetab.py; this was originally added in f2c4cedf.

andrewcarlotti commented:
I'm not sure what went wrong in some of those tests.
Firstly, the known:
- The archlinux webkit failure and the py36-pyqt59 tests both failed due to a race condition in whether a history entry is marked as redirected before or after the new page sets the title.
- py38-pyqt514-cov failed due to incomplete code coverage - specifically in the case where the same redirected link is opened in two tabs simultaneously. I'll write a unit test for this (an end2end test might miss it if the test spans a change of second).

Now one I don't understand:
https://ci.appveyor.com/project/qutebrowser/qutebrowser/builds/32049893/job/wnkv0lofqrpf8yr7
The VS 2019 test failed due to a timeout while waiting for "data/backforward/3.txt" to load. In line 381 it seems to be loading 3.txt, but subsequent lines seem to suggest that this somehow switches back to having successfully loaded 2.txt. I've now idea how this could happen, and I don't see how it would be affected my the changes I made.

Finally, there are a lot of failures with the py35-pyqt57 configuration, which seem to be a mixture of SQL errors with the UPDATE statement, and timeout errors waiting for page loads.

andrewcarlotti commented:
I think I've fixed all the test failures that I understand. There's still the issue with the sql update statement - I don't know why it's not accepted on the oldest configuration. Besides this, there are the strange timeouts on page loads, and I'm now also getting reproducible failures with the tests/unit/browser/test_caret.py:TestSearch tests - I can also reproduce these on the current master.

So overall, I think the issue with the sql query still needs investigating, but I'm not sure where to go with that. The other test failures aren't obviously related to anything I've done, but since i don't understand them I can't entirely rule that out.

rcorre commented:
That test box is using an older OS (Ubuntu 16.04.6 LTS), so its possible you're using a newer feature of sqlite3, though I'm not sure what.

rcorre commented:
That version of ubuntu uses `sqlite: 3.11.1` (3.31.1 is the most recent release)

The-Compiler commented:
> and I'm now also getting reproducible failures with the tests/unit/browser/test_caret.py:TestSearch tests - I can also reproduce these on the current master

You mean you can also reproduce those locally? What are you running exactly to do so? They run fine for me.

andrewcarlotti commented:
> > and I'm now also getting reproducible failures with the tests/unit/browser/test_caret.py:TestSearch tests - I can also reproduce these on the current master
> 
> You mean you can also reproduce those locally? What are you running exactly to do so? They run fine for me.

I just created a fresh clone of the repository, and then ran the tests. I misremembered slightly - it fails during the same tests as the failure on Travis, but with more output. I've copied this here:
```
$ tox -e py38-pyqt514-cov tests/unit/browser/test_caret.py
py38-pyqt514-cov create: {REPO_DIR}/.tox/py38-pyqt514-cov
py38-pyqt514-cov installdeps: -r{REPO_DIR}/requirements.txt, -r{REPO_DIR}/misc/requirements/requirements-tests.txt, -r{REPO_DIR}/misc/requirements/requirements-pyqt-5.14.txt
py38-pyqt514-cov installed: attrs==19.3.0,beautifulsoup4==4.8.2,cheroot==8.3.0,click==7.1.1,colorama==0.4.3,coverage==5.0.4,cssutils==1.0.2,EasyProcess==0.2.10,Flask==1.1.1,glob2==0.7,hunter==3.1.3,hypothesis==5.8.0,itsdangerous==1.1.0,jaraco.functools==3.0.0,Jinja2==2.11.1,Mako==1.1.2,manhole==1.6.0,MarkupSafe==1.1.1,more-itertools==8.2.0,packaging==20.3,parse==1.15.0,parse-type==0.5.2,pluggy==0.13.1,py==1.8.1,py-cpuinfo==5.0.0,Pygments==2.6.1,pyparsing==2.4.6,pyPEG2==2.15.2,PyQt5==5.14.2,PyQt5-sip==12.7.2,PyQtWebEngine==5.14.0,pytest==5.4.1,pytest-bdd==3.2.1,pytest-benchmark==3.2.3,pytest-cov==2.8.1,pytest-instafail==0.4.1.post0,pytest-mock==2.0.0,pytest-qt==3.3.0,pytest-repeat==0.8.0,pytest-rerunfailures==9.0,pytest-travis-fold==1.3.0,pytest-xvfb==1.2.0,PyVirtualDisplay==0.2.5,PyYAML==5.3.1,six==1.14.0,sortedcontainers==2.1.0,soupsieve==2.0,vulture==1.4,wcwidth==0.1.9,Werkzeug==1.0.0
py38-pyqt514-cov run-test-pre: PYTHONHASHSEED='2868701338'
py38-pyqt514-cov run-test: commands[0] | {REPO_DIR}/.tox/py38-pyqt514-cov/bin/python scripts/link_pyqt.py --tox {REPO_DIR}/.tox/py38-pyqt514-cov
LINK_PYQT_SKIP set, exiting...
py38-pyqt514-cov run-test: commands[1] | {REPO_DIR}/.tox/py38-pyqt514-cov/bin/python -bb -m pytest tests/unit/browser/test_caret.py
pytest-xvfb could not find Xvfb. You can install it to prevent windows from being shown.
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
cachedir: .tox/py38-pyqt514-cov/.pytest_cache
PyQt5 5.14.2 -- Qt runtime 5.14.2 -- Qt compiled 5.14.2
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: {REPO_DIR}, inifile: pytest.ini
plugins: cov-2.8.1, mock-2.0.0, hypothesis-5.8.0, qt-3.3.0, rerunfailures-9.0, xvfb-1.2.0, repeat-0.8.0, instafail-0.4.1.post0, travis-fold-1.3.0, benchmark-3.2.3, bdd-3.2.1
collected 78 items

tests/unit/browser/test_caret.py s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.sF

_______________________________________________________________________________ TestSearch.test_yanking_a_searched_line[webengine] ________________________________________________________________________________

self = <test_caret.TestSearch object at 0x7f13ec5179d0>, caret = <qutebrowser.browser.webengine.webenginetab.WebEngineCaret object at 0x7f13ccef6940>, selection = <test_caret.Selection object at 0x7f13cce50550>
mode_manager = <qutebrowser.keyinput.modeman.ModeManager mode=<KeyMode.normal: 1>>, web_tab = <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=28 url='qute://testdata/data/caret.html'>
qtbot = <pytestqt.qtbot.QtBot object at 0x7f13ccf4c550>

    @pytest.mark.qtbug60673
    @pytest.mark.no_xvfb
    def test_yanking_a_searched_line(self, caret, selection, mode_manager, web_tab, qtbot):
        mode_manager.leave(usertypes.KeyMode.caret)
    
        with qtbot.wait_callback() as callback:
            web_tab.search.search('fiv', result_cb=callback)
>       callback.assert_called_with(True)

tests/unit/browser/test_caret.py:310: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pytestqt.wait_signal.CallbackBlocker object at 0x7f13cce508b0>, args = (True,), kwargs = {}, @py_assert1 = [False], @py_assert6 = [True], @py_assert3 = False
@py_format8 = '[False] == [True]\n~At index 0 diff: False != True\n~Use -v to get the full diff', @py_format10 = 'assert [False] == [True]\n~At index 0 diff: False != True\n~Use -v to get the full diff'

    def assert_called_with(self, *args, **kwargs):
        """
        Check that the callback was called with the same arguments as this
        function.
        """
        assert self.called
>       assert self.args == list(args)
E       assert [False] == [True]
E         At index 0 diff: False != True
E         Use -v to get the full diff

.tox/py38-pyqt514-cov/lib/python3.8/site-packages/pytestqt/wait_signal.py:681: AssertionError
----------------------------------------------------------------------------------------------- Captured log setup ------------------------------------------------------------------------------------------------
DEBUG    greasemonkey:greasemonkey.py:284 Reading scripts from: /tmp/pytest-of-andrew/pytest-16/test_yanking_a_searched_line_w1/data/greasemonkey
DEBUG    config:config.py:323 Config option changed: content.javascript.log = {'info': 'info', 'error': 'error', 'unknown': 'error', 'warning': 'error'}
DEBUG    webview:browsertab.py:961 Going to start loading: qute://testdata/data/caret.html
DEBUG    webview:browsertab.py:961 Going to start loading: qute://testdata/data/caret.html
DEBUG    misc:eventfilter.py:55 <qutebrowser.browser.webengine.webview.WebEngineView object at 0x7f13ccef6700> got new child <PyQt5.QtWidgets.QWidget object at 0x7f13ccef6ee0>, installing filter
DEBUG    destroy:objreg.py:125 schedule removal: 27
DEBUG    destroy:objreg.py:125 schedule removal: tab
DEBUG    misc:eventfilter.py:80 <PyQt5.QtWidgets.QWidget object at 0x7f13ccef6e50>: removed child <PyQt5.QtCore.QObject object at 0x7f13ccef9040>
DEBUG    misc:eventfilter.py:80 <PyQt5.QtWidgets.QWidget object at 0x7f13ccef6e50>: removed child <PyQt5.QtWidgets.QWidget object at 0x7f13ccca6d30>
DEBUG    destroy:objreg.py:130 removed: 27
DEBUG    destroy:objreg.py:130 removed: tab
DEBUG    webview:browsertab.py:927 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=28 url='qute://testdata/data/caret.html'>: LoadStatus.loading
DEBUG    webview:browsertab.py:986 navigation request: url qute://testdata/data/caret.html, type Type.typed, is_main_frame True
DEBUG    misc:webenginequtescheme.py:122 Got request for qute://testdata/data/caret.html
DEBUG    misc:qutescheme.py:146 url: qute://testdata/data/caret.html, path: /data/caret.html, host testdata
DEBUG    misc:webenginequtescheme.py:146 Returning text/html data
DEBUG    webview:browsertab.py:927 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=28 url='qute://testdata/data/caret.html'>: LoadStatus.success
DEBUG    modes:modeman.py:317 Entering mode KeyMode.caret
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
DEBUG    modes:modeman.py:384 Leaving mode KeyMode.caret
DEBUG    webview:webenginetab.py:280 Ignoring selection cb due to mode change.
DEBUG    webview:webenginetab.py:203 search didn't find fiv with flags FindCaseSensitively

tests/unit/browser/test_caret.py sF

____________________________________________________________________ TestSearch.test_yanking_a_searched_line_with_multiple_matches[webengine] _____________________________________________________________________

self = <test_caret.TestSearch object at 0x7f13ccef4580>, caret = <qutebrowser.browser.webengine.webenginetab.WebEngineCaret object at 0x7f13ccacd0d0>, selection = <test_caret.Selection object at 0x7f13ccb96700>
mode_manager = <qutebrowser.keyinput.modeman.ModeManager mode=<KeyMode.normal: 1>>, web_tab = <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=29 url='qute://testdata/data/caret.html'>
qtbot = <pytestqt.qtbot.QtBot object at 0x7f13ccafcc40>

    @pytest.mark.qtbug60673
    @pytest.mark.no_xvfb
    def test_yanking_a_searched_line_with_multiple_matches(self, caret, selection, mode_manager, web_tab, qtbot):
        mode_manager.leave(usertypes.KeyMode.caret)
    
        with qtbot.wait_callback() as callback:
            web_tab.search.search('w', result_cb=callback)
>       callback.assert_called_with(True)

tests/unit/browser/test_caret.py:323: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pytestqt.wait_signal.CallbackBlocker object at 0x7f13ccb968e0>, args = (True,), kwargs = {}, @py_assert1 = [False], @py_assert6 = [True], @py_assert3 = False
@py_format8 = '[False] == [True]\n~At index 0 diff: False != True\n~Use -v to get the full diff', @py_format10 = 'assert [False] == [True]\n~At index 0 diff: False != True\n~Use -v to get the full diff'

    def assert_called_with(self, *args, **kwargs):
        """
        Check that the callback was called with the same arguments as this
        function.
        """
        assert self.called
>       assert self.args == list(args)
E       assert [False] == [True]
E         At index 0 diff: False != True
E         Use -v to get the full diff

.tox/py38-pyqt514-cov/lib/python3.8/site-packages/pytestqt/wait_signal.py:681: AssertionError
----------------------------------------------------------------------------------------------- Captured log setup ------------------------------------------------------------------------------------------------
DEBUG    greasemonkey:greasemonkey.py:284 Reading scripts from: /tmp/pytest-of-andrew/pytest-16/test_yanking_a_searched_line_w3/data/greasemonkey
DEBUG    config:config.py:323 Config option changed: content.javascript.log = {'info': 'info', 'error': 'error', 'unknown': 'error', 'warning': 'error'}
DEBUG    webview:browsertab.py:961 Going to start loading: qute://testdata/data/caret.html
DEBUG    webview:browsertab.py:961 Going to start loading: qute://testdata/data/caret.html
DEBUG    misc:eventfilter.py:55 <qutebrowser.browser.webengine.webview.WebEngineView object at 0x7f13ccb92e50> got new child <PyQt5.QtWidgets.QWidget object at 0x7f13ccacd670>, installing filter
DEBUG    destroy:objreg.py:125 schedule removal: 28
DEBUG    destroy:objreg.py:125 schedule removal: tab
DEBUG    misc:eventfilter.py:80 <PyQt5.QtWidgets.QWidget object at 0x7f13ccacd5e0>: removed child <PyQt5.QtCore.QObject object at 0x7f13ccacd790>
DEBUG    misc:eventfilter.py:80 <PyQt5.QtWidgets.QWidget object at 0x7f13ccacd5e0>: removed child <PyQt5.QtWidgets.QWidget object at 0x7f13ccef6ee0>
DEBUG    destroy:objreg.py:130 removed: 28
DEBUG    destroy:objreg.py:130 removed: tab
DEBUG    webview:browsertab.py:927 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=29 url='qute://testdata/data/caret.html'>: LoadStatus.loading
DEBUG    webview:browsertab.py:986 navigation request: url qute://testdata/data/caret.html, type Type.typed, is_main_frame True
DEBUG    misc:webenginequtescheme.py:122 Got request for qute://testdata/data/caret.html
DEBUG    misc:qutescheme.py:146 url: qute://testdata/data/caret.html, path: /data/caret.html, host testdata
DEBUG    misc:webenginequtescheme.py:146 Returning text/html data
DEBUG    webview:browsertab.py:927 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=29 url='qute://testdata/data/caret.html'>: LoadStatus.success
DEBUG    modes:modeman.py:317 Entering mode KeyMode.caret
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
DEBUG    modes:modeman.py:384 Leaving mode KeyMode.caret
DEBUG    webview:webenginetab.py:280 Ignoring selection cb due to mode change.
DEBUG    webview:webenginetab.py:203 search didn't find w with flags FindCaseSensitively

tests/unit/browser/test_caret.py s.s.s.s.s.s.s.s.s.                                                                                                                                                         [100%]

----------- coverage: platform linux, python 3.8.2-final-0 -----------
Coverage HTML written to dir htmlcov
Coverage XML written to file coverage.xml


============================================================================================= short test summary info =============================================================================================
FAILED tests/unit/browser/test_caret.py::TestSearch::test_yanking_a_searched_line[webengine] - assert [False] == [True]
FAILED tests/unit/browser/test_caret.py::TestSearch::test_yanking_a_searched_line_with_multiple_matches[webengine] - assert [False] == [True]
==================================================================================== 2 failed, 37 passed, 39 skipped in 40.39s ====================================================================================
Fatal Python error: Segmentation fault

Current thread 0x00007f14423d7740 (most recent call first):
<no Python frame>
ERROR: InvocationError for command {REPO_DIR}/.tox/py38-pyqt514-cov/bin/python -bb -m pytest tests/unit/browser/test_caret.py (exited with code -11 (SIGSEGV)) (exited with code -11)
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
ERROR:   py38-pyqt514-cov: commands failed
```

andrewcarlotti commented:
> That version of ubuntu uses `sqlite: 3.11.1` (3.31.1 is the most recent release)

From https://www.sqlite.org/rowvalue.html:
"Row values were added to SQLite version 3.15.0 (2016-10-14). Attempts to use row values in prior versions of SQLite will generate syntax errors."

I'll rewrite the other half of the statement too then.

andrewcarlotti commented:
I'm happy with this now. I think the only outstanding question is whether/when reloading a page should add new history items, but that can be answered/added later. The latest set of tests all passed, but I think there's some existing issues (i.e. unrelated to my changes) with some of them leading to occasional strange failures.

toofar commented on qutebrowser/browser/browsertab.py:
```diff
@@ -1041,9 +1052,83 @@ def _update_load_status(self, ok: bool) -> None:
+            # Update this in case it changed.
+            self._last_history_requested_url = requested_url
+        elif force_entry or title != self._last_history_title:
+            # Don't add an entry if only the url has changed.
```
How about the case of navigating to two pages where the title is generated from the URL? Perhaps this should be something like `elif force_entry or not title or title != self._last_history_title:`?

andrewcarlotti replied:
Oh, good point. I'll change that.

andrewcarlotti replied:
On second thoughts: the title check is only for the case where the url has changed due to a javascript trigger (or something similar). Pages that are loaded 'normally' will have a new history entry generated by the trigger being called from a loadFinished signal (or the WebKit equivalent). So the only case this would affect is sites triggering Javascript/HTML5 url changes that haven't set a title, and I think it's reasonable to assume it's still the same page in this case (at least if we're doing a similar thing for an unchanged non-empty title).

toofar commented on tests/unit/browser/test_history.py:
```diff
@@ -205,60 +205,76 @@ def raise_error(url, replace=False):
         assert list(web_history)
         assert not list(web_history.completion)
 
-    def test_no_immedate_duplicates(self, web_history, mock_time):
```
Why remove this test, you are preserving this behaviour currently right?

andrewcarlotti replied:
I'm changing the behaviour slightly to operate on a per-tab basis. The existing test assumes that deduplication happens globally within the WebHistory class.

andrewcarlotti commented:
> 2\. Should reloading a page create a new entry? Currently the behaviour depends on whether there have been any page loads in other tabs. **With this change, reloading would never create a new entry.** There's an argument to be made that it should at least sometimes create a new entry - either at all times, or whenever the url has changed since the last history entry (from that tab). Either of these would help mitigate the issue in the first question.

I was wrong here - reloading will lead to `_on_history_trigger` being called with `force_entry=True`, so if the URL has changed, then reloading will create a new history entry. I think this is probably the best choice of behaviour.

andrewcarlotti commented:
An example of a site that still misbehaves is www.bbc.co.uk/iplayer (probably not sufficiently accessible outside the UK). When I click to go to individual programme pages, the resulting navigation request has type 'redirect' (for comparison, YouTube triggers requests with type 'other').

This is then made worse by the completion history potentially becoming 'out of sync' with history - if the page exists earlier in history (without a redirect), then it ideally should still be present in the completion history. Unfortuantely, with my current changes, this old instance is updated prior to the redirect, and then deleted when the redirect happens. This issue could also be triggered by things like captive portals, web filters and login pages.

I think the solution to this requires searching history when deleting a completion history entry to see if it should be restored at an older time. (I'll implement this later). It also might affect the best choice of behaviour upon reloading a page - if we load page X, then reload page X, then page X redirects to page Y, should the (completion) history display both page X and page Y, or just page Y?

andrewcarlotti commented:
Once again, the only test failures appear to be flaky tests - all the changes I've made seem to test fine. I've also been running my latest versions for a couple of days now, which has improved the quality of my browsing history with no obvious (unfixed) problems.

andrewcarlotti commented:
I found another mishandled edge case, which I observed when loading https://www.facebook.com/pg/{PAGE}/community/. Facebook rewrites the URL to remove the "pg" component, then sets the title, then rewrites sets the URL back to the original. Prior to my latest commit, this would create two visible history entries (i.e. both would have redirect=1); I've now fixed that.

Hi-Angel commented:
> Whenever the page title changes and the page url does not, then the new title is saved in the history entry in place of the old one

Unless I am missing something, this sounds very wasteful. Some sites change the page title for no serious reason. For example https://vk.com/ *(a social network mainly popular in slavic countries)* makes page title blink with 1-second frequency by changing its current title to a text "X new messages" whenever a new message comes and you didn't open the tab.

andrewcarlotti commented:
> Unless I am missing something, this sounds very wasteful. Some sites change the page title for no serious reason. For example https://vk.com/ _(a social network mainly popular in slavic countries)_ makes page title blink with 1-second frequency by changing its current title to a text "X new messages" whenever a new message comes and you didn't open the tab.

You make a good point. That behaviour was intended to handle the case where the page title is set after page is loaded, but setting it on every title change does seem excessive. What would seem desirable is to apply some sort of 'debouncing' to the updates, but with the current architecture (where the code is only run when certain events happen) I'm not yet sure how that could work.

The-Compiler commented:
> What would seem desirable is to apply some sort of 'debouncing' to the updates, but with the current architecture (where the code is only run when certain events happen) I'm not yet sure how that could work.

Perhaps something like `misc.throttle.Throttle`?

(Also, sorry for the radio silence from my side - I really want to take a first look at this,  but at usual, my backlog is rather big.)

andrewcarlotti commented:
I've rebased my original work on top of the latest master (and squashed it to avoid having to rehandle several conflicts I had previously resolved in a merge).

I haven't implemented any change to throttle title updates yet. I think the best idea I have is to detect titles that cycle through repeated states is the following:

Let n=0, k=0
Let best_title=None
Let best_title_count=0
For each title update:
- Update the database
- If new_title is the better than best_title, then set best_title=new_title
- If new_title==best_title, then increment best_title_count
- If best_title_count>=3, then stop and ignore any future title updates
- Increment n
- If n==2^k, then set n=0, increment k, and reset best_title=None and best_title_count=0

The idea being that if the title continues alternating between a few different values, then we eventually detect this and leave the "best" recurring title saved in the database. I think "best" should mean "shortest" here, but I'm not totally sure of this - there may be reasons to pick the longest title instead.

As described, the algorithm would only stop title updates changing the database after a minimum of 12 title changes.

Hi-Angel commented:
Unrelated to the algorithm suggestion, I have two more problems I think are important to consider.

First is "what is the correct page title for sites, where page title can be changed on event *(such as a new message)*?

Github is a nice example here. If you have a github issue/pull tab opened, then once a new comment appears, the tab name changes to something like `[1] Page title follows`. Obviously, if you save this title into DB, this would be wrong, because the correct title is `Page title follows`, without the `[1] ` that signifies a new message.

---------

Another problem. Let's consider a site that this is supposed to fix. Let's say, this site has URL https://example.com and two pages "Main page" and "Inbox". When you click "Inbox" link, the URL is still https://example.com because it's one of those terrible sites that doesn't change their URL but only the title.

Okay, so here's the thing: next time I want to go to this site, I get a completion `https://example.com "Inbox"`. Does that mean this completion supposed to take me to my Inbox? No, it will take me to "Main page" instead, which is confusing.

---------

I think these problem either need to be taken into account somehow, or I'd suggest the behavior this PR does should be optional *(with whether opt-in or opt-out being something to debate)*, because while this would improve experience for one users, but at the same time it would regress it for others *(for example, I use completions everyday but history — here as opposed to completions — almost never. So I care the most about the completions correctness rather than history correctness)*.

andrewcarlotti commented:
> First is "what is the correct page title for sites, where page title can be changed on event _(such as a new message)_?
> 
> Github is a nice example here. If you have a github issue/pull tab opened, then once a new comment appears, the tab name changes to something like `[1] Page title follows`. Obviously, if you save this title into DB, this would be wrong, because the correct title is `Page title follows`, without the `[1] ` that signifies a new message.

It might be strictly wrong compared to the an ideal choice, but in this example it is wrong in a way that is easy to mentally correct when reading it.

> I think these problem either need to be taken into account somehow,

I don't think this is fully possible, but I'm certainly happy to hear about examples that might break one heuristic or another to help inform better heuristics. Ultimately there is no perfect heuristic - all we can do is try to get the right answer, or something close to the right answer, as often as possible.

> or I'd suggest the behavior this PR does should be optional _(with whether opt-in or opt-out being something to debate)_, because while this would improve experience for one users, but at the same time it would regress it for others _(for example, I use completions everyday but history — here as opposed to completions — almost never. So I care the most about the completions correctness rather than history correctness)_.

I don't think there is a difference between history correctness and completion history correctness, since they use almost exactly the same data (completion history takes the full history and filters out redirects and duplicate urls). Instead, it is a case of trying to find behaviour that is correct as often as possible - your issue is that for a specific site you care about the status quo works better, whereas I think there are many more sites where the heuristic in this pull request would work better.

----------------------------------------------


My motivating example for doing this was that most of my YouTube browsing history was present with the wrong title (and often not present at all), because the title is set by Javascript after the initial load is complete. 100s of entries in my completion history with an inscrutable url and the title "YouTube" is approximately useless. This behaviour is also present on many other sites, which is why I think just saving the initial title is not the best default behaviour.

Unfortunately identifying the 'correct' canonical title is difficult. Consider the example of opening a Facebook page when you have an unread message. The title looks something like "Alex messaged you". You then read the message and the title is set back to something more relevant to the page you loaded. Compare that with your example:

> Let's consider a site that this is supposed to fix. Let's say, this site has URL https://example.com and two pages "Main page" and "Inbox". When you click "Inbox" link, the URL is still https://example.com because it's one of those terrible sites that doesn't change their URL but only the title.
> 
> Okay, so here's the thing: next time I want to go to this site, I get a completion `https://example.com "Inbox"`. Does that mean this completion supposed to take me to my Inbox? No, it will take me to "Main page" instead, which is confusing.

As far as the browser can tell, I think these two examples include roughly the same sequence of user interaction and url and title updates. However, in the first example the second title is clearly more 'canonical', whereas in the second example the first title is more canonical.

I don't currently know of a way to completely solve this without adding scope for site-specific rules, but then you'd almost be handwriting parts of your history. I do, however, think that the heuristic in this pull request is better than the status quo. I'd happily take a confusing title for a root page over a generic title for a specific content page with an inscrutable url - in the first scenario I can just use the url for searching completions (and 'autocorrect' the title in my mind), whereas the only remedy in the second scenario is to revisit every page individually (which is completely impractical for 100s or 1000s of YouTube videos).

toofar commented:
Hey @andrewcarlotti, I've attempted to rebase on the latest main branch with Qt6. I force pushed to your master branch (lol, didn't actually expect that to work, probably should have pushed a merge commit), hope that's okay!

I didn't attempt to get the linters passing, I think at least mypy will be complaining.
Also not sure it's working the same between webkit and webengine (I'm testing with `python3 -m qutebrowser -T --backend webkit --qt-wrapper PyQt5 https://goswagger.io/ qute://history`). Webengine seems to get a new entry for each page on that SPA, but for webkit it seems to update the old entry. Not sure if I broke it or it was always like that.

Anyway, FWIW I've been using this for a few years now and while I think there are potential a few behaviours that could use polishing I think it is better than the implementation on main.
